### PR TITLE
Refactor readers to reduce surface area

### LINKED ISSUE #1917: v2 - We should NOT do sync over async
6 instances being suppressed 

https://github.com/search?q=repo%3Amicrosoft%2FOpenAPI.NET%20VSTHRD002%20&type=code

This will cause poor performance and potential dead locks on backend services.

One of those instances is marked as obsolete, we should simply clean it up (and make a pass to ensure we remove any other obsolete API)

The five others were introduced with this preview. We should remove those methods, and let the consumer do the blocking if they absolutely need to with the proper patterns in terms of thread and context management

### LINKED ISSUE #1918: v2 - loading the document requires knowing the format in advance
```chsarp
new OpenApiStreamReader().Read(stream, out var diagnostic);
```

(or its async variant)

Did not require knowing the format of the document (JSON/YAML) in advance. This was handy as when building general purpose tools, it's not something that I know in advance. Parsing the file extension/url extension/response content type header _might_ help discovering it, but it's not reliable.

The equivalent method

```csharp
OpenApiDocument.LoadAsync(input, "unknown", settings)
```

Requires passing the format, which I believe is a regression in functionality. We should work to restore the same level of functionality.

Also, there are a number of places in the documentation and doc comments that still refer to the old initialization and require to be updated before GA.

https://github.com/search?q=repo%3Amicrosoft%2FOpenAPI.NET%20OpenApiStreamReader&type=code

### LINKED ISSUE #1951: Calling OpenApiDocument.Parse without the `format` parameter should not fail
**Describe the bug**

**Code To Reproduce**

```csharp

    OpenApiReaderRegistry.RegisterReader("yaml",new OpenApiYamlReader());

    var stringOpenApiDoc = """
openapi: 3.1.0
info:
  title: Sample API
  version: 1.0.0
paths: {}
""";

    var readResult = OpenApiDocument.Parse(stringOpenApiDoc);
```

**Expected behavior**
Document should parse without error.

**Screenshots/Code Snippets**

Error message returned
```shell
OpenAPI document is invalid:
Please provide the correct format, 'o' is an invalid start of a value. LineNumber: 0 | BytePositionInLine: 0.
```

### LINKED ISSUE #1954: Remove Load method that accepts a Url
We should remove this overload.  Accessing a URL is an asynchronous process, if a caller wants to hide it behind a sync call, then that is their responsibility.

https://github.com/microsoft/OpenAPI.NET/blob/1338905859eaa083f89a8819227073feef8836a3/src/Microsoft.OpenApi/Reader/OpenApiModelFactory.cs#L34

### LINKED ISSUE #1964: [API review] Evaluate format default value in Load methods
Where the loading methods are used without a format property, we will attempt to parse the document using the default JSON reader. If that fails and the Yaml reader is registered, then we will attempt to read as YAML. 

Evaluate if the following methods should have default format value of "json" or the best way to manage the default format value: 
``` 
public class OpenApiDocument {
    public static async Task<ReadResult> LoadAsync(string url, OpenApiReaderSettings settings = null) {}
    public static async Task<ReadResult> LoadAsync(Stream stream, string? format = null, OpenApiReaderSettings? settings = null) {}
    public static ReadResult Load(MemoryStream stream, string? format = null, OpenApiReaderSettings? settings = null) {}
    public static ReadResult Parse(string input,string? format = null,OpenApiReaderSettings? settings = null) {}
}
```

### PR DESCRIPTION
**Pull request by MaggieKimani1**
Fixes #1954
fixes #1951
Fixes #1964
Fixes #1917
fixes #1918
closes #1958
closes #1929

### CODE REVIEW

File: src/Microsoft.OpenApi/Reader/OpenApiModelFactory.cs [hunk 1]
```diff
@@ -153,29 +170,13 @@ public static ReadResult Parse(string input,
...
+            format ??= InspectInputFormat(input);
             settings ??= new OpenApiReaderSettings();
-            using var reader = new StringReader(input);
 
-#pragma warning disable VSTHRD002 // Avoid problematic synchronous waits
-            return ParseAsync(input, reader, format, settings).GetAwaiter().GetResult();
-#pragma warning restore VSTHRD002 // Avoid problematic synchronous waits
-        }
+            // Copy string into MemoryStream
+            var stream = new MemoryStream(Encoding.UTF8.GetBytes(input));
```
Comment (baywet) [inline src/Microsoft.OpenApi/Reader/OpenApiModelFactory.cs:177]:
```suggestion
            using var stream = new MemoryStream(Encoding.UTF8.GetBytes(input));
```

File: src/Microsoft.OpenApi/Reader/OpenApiModelFactory.cs [hunk 2]
```diff
@@ -153,29 +170,13 @@ public static ReadResult Parse(string input,
                                        string format = null,
                                        OpenApiReaderSettings settings = null)
         {
-            format ??= OpenApiConstants.Json;
+            format ??= InspectInputFormat(input);
```
Comment (baywet) [inline src/Microsoft.OpenApi/Reader/OpenApiModelFactory.cs:173]: defensive programming

File: src/Microsoft.OpenApi/Reader/OpenApiModelFactory.cs [hunk 3]
```diff
@@ -193,146 +194,172 @@ public static T Parse<T>(string input,
                                  string format = null,
                                  OpenApiReaderSettings settings = null) where T : IOpenApiElement
         {
-            format ??= OpenApiConstants.Json;
+            format ??= InspectInputFormat(input);
             settings ??= new OpenApiReaderSettings();
-            using var reader = new StringReader(input);
-            return Load<T>(reader, version, out diagnostic, format, settings);
+            var stream = new MemoryStream(Encoding.UTF8.GetBytes(input));
```
Comment (baywet) [inline src/Microsoft.OpenApi/Reader/OpenApiModelFactory.cs:199]:
```suggestion
            using var stream = new MemoryStream(Encoding.UTF8.GetBytes(input));
```

File: src/Microsoft.OpenApi/Reader/OpenApiModelFactory.cs [hunk 4]
```diff
@@ -193,146 +194,172 @@ public static T Parse<T>(string input,
                                  string format = null,
                                  OpenApiReaderSettings settings = null) where T : IOpenApiElement
         {
-            format ??= OpenApiConstants.Json;
+            format ??= InspectInputFormat(input);
```
Comment (baywet) [inline src/Microsoft.OpenApi/Reader/OpenApiModelFactory.cs:197]: defensive programming

File: src/Microsoft.OpenApi/Reader/OpenApiModelFactory.cs [hunk 5]
```diff
@@ -193,146 +194,172 @@ public static T Parse<T>(string input,
...
-        /// <param name="settings">The OpenApiReader settings.</param>
-        /// <returns>Instance of newly created IOpenApiElement.</returns>
-        /// <returns>The OpenAPI element.</returns>
-        public static T Load<T>(string url, OpenApiSpecVersion version, out OpenApiDiagnostic diagnostic, OpenApiReaderSettings settings = null) where T : IOpenApiElement
+        private static async Task<ReadResult> InternalLoadAsync(Stream input, string format, OpenApiReaderSettings settings, CancellationToken cancellationToken = default)
         {
-            var format = GetFormat(url);
-            settings ??= new OpenApiReaderSettings();
+            var reader = OpenApiReaderRegistry.GetReader(format);
+            var readResult = await reader.ReadAsync(input, settings, cancellationToken);
```
Comment (baywet) [inline src/Microsoft.OpenApi/Reader/OpenApiModelFactory.cs:206]:
```suggestion
            var readResult = await reader.ReadAsync(input, settings, cancellationToken).ConfigureAwait(false);
```

File: src/Microsoft.OpenApi/Reader/OpenApiModelFactory.cs [hunk 6]
```diff
@@ -193,146 +194,172 @@ public static T Parse<T>(string input,
...
 
-#pragma warning disable VSTHRD002 // Avoid problematic synchronous waits
-            var stream = GetStreamAsync(url).GetAwaiter().GetResult();
-#pragma warning restore VSTHRD002 // Avoid problematic synchronous waits
+            if (settings.LoadExternalRefs)
+            {
+                var diagnosticExternalRefs = await LoadExternalRefsAsync(readResult.OpenApiDocument, cancellationToken, settings, format);
```
Comment (baywet) [inline src/Microsoft.OpenApi/Reader/OpenApiModelFactory.cs:210]:
```suggestion
                var diagnosticExternalRefs = await LoadExternalRefsAsync(readResult.OpenApiDocument, cancellationToken, settings, format).ConfigureAwait(false);
```

File: src/Microsoft.OpenApi/Reader/OpenApiModelFactory.cs [hunk 7]
```diff
@@ -193,146 +194,172 @@ public static T Parse<T>(string input,
...
-        /// <typeparam name="T"></typeparam>
-        /// <param name="input">Stream containing OpenAPI description to parse.</param>
-        /// <param name="version">Version of the OpenAPI specification that the fragment conforms to.</param>
-        /// <param name="format"></param>
-        /// <param name="diagnostic">Returns diagnostic object containing errors detected during parsing.</param>
-        /// <param name="settings">The OpenApiReader settings.</param>
-        /// <returns>Instance of newly created IOpenApiElement.</returns>
-        /// <returns>The OpenAPI element.</returns>
-        public static T Load<T>(Stream input, OpenApiSpecVersion version, string format, out OpenApiDiagnostic diagnostic, OpenApiReaderSettings settings = null) where T : IOpenApiElement
+        private static async Task<OpenApiDiagnostic> LoadExternalRefsAsync(OpenApiDocument document, CancellationToken cancellationToken, OpenApiReaderSettings settings, string format = null)
```
Comment (baywet) [inline src/Microsoft.OpenApi/Reader/OpenApiModelFactory.cs:222]: nit: move cancellation token parameter to the last, make it optional to follow conventions

File: src/Microsoft.OpenApi/Reader/OpenApiModelFactory.cs [hunk 8]
```diff
@@ -193,146 +194,172 @@ public static T Parse<T>(string input,
...
-            using var reader = new StreamReader(input);
-            return Load<T>(reader, version, out diagnostic, format, settings);
+            // Create workspace for all documents to live in.
+            var baseUrl = settings.BaseUrl ?? new Uri(OpenApiConstants.BaseRegistryUri);
+            var openApiWorkSpace = new OpenApiWorkspace(baseUrl);
+
+            // Load this root document into the workspace
+            var streamLoader = new DefaultStreamLoader(settings.BaseUrl);
+            var workspaceLoader = new OpenApiWorkspaceLoader(openApiWorkSpace, settings.CustomExternalLoader ?? streamLoader, settings);
+            return await workspaceLoader.LoadAsync(new OpenApiReference() { ExternalResource = "/" }, document, format ?? OpenApiConstants.Json, null, cancellationToken);
```
Comment (baywet) [inline src/Microsoft.OpenApi/Reader/OpenApiModelFactory.cs:231]:
```suggestion
            return await workspaceLoader.LoadAsync(new OpenApiReference() { ExternalResource = "/" }, document, format ?? OpenApiConstants.Json, null, cancellationToken).ConfigureAwait(false);
```

File: src/Microsoft.OpenApi/Reader/OpenApiModelFactory.cs [hunk 9]
```diff
@@ -193,146 +194,172 @@ public static T Parse<T>(string input,
...
-                    var path = new Uri(url);
-                    var urlSuffix = path.Segments[path.Segments.Length - 1].Split('.').LastOrDefault();
+                Stream stream;
+                string format;
 
-                    return !string.IsNullOrEmpty(urlSuffix) ? urlSuffix : GetContentType(url).Split('/').LastOrDefault();
+                if (url.StartsWith("http", StringComparison.OrdinalIgnoreCase)
+                    || url.StartsWith("https", StringComparison.OrdinalIgnoreCase))
+                {
+                    var response = await _httpClient.GetAsync(url);
```
Comment (baywet) [inline src/Microsoft.OpenApi/Reader/OpenApiModelFactory.cs:257]: conditional compilation and pass cancellation token.
Also configure await

File: src/Microsoft.OpenApi/Reader/OpenApiModelFactory.cs [hunk 10]
```diff
@@ -193,146 +194,172 @@ public static T Parse<T>(string input,
...
+                    var mediaType = response.Content.Headers.ContentType.MediaType;
+                    var contentType = mediaType.Split(";".ToCharArray(), StringSplitOptions.RemoveEmptyEntries)[0];
+                    format = contentType.Split('/').LastOrDefault();
+                    stream = await response.Content.ReadAsStreamAsync();
```
Comment (baywet) [inline src/Microsoft.OpenApi/Reader/OpenApiModelFactory.cs:261]: conditional compilation for token, also configure await

File: src/Microsoft.OpenApi/Reader/OpenApiModelFactory.cs [hunk 11]
```diff
@@ -193,146 +194,172 @@ public static T Parse<T>(string input,
...
+        private static async Task<(Stream, string)> PrepareStreamForReadingAsync(Stream input, string format, CancellationToken token = default)
+        {
+            Stream preparedStream = input;
+
+            if (!input.CanSeek)
             {
-                try
+                // Use a temporary buffer to read a small portion for format detection
+                using var bufferStream = new MemoryStream();
+                await input.CopyToAsync(bufferStream, 1024, token);
```
Comment (baywet) [inline src/Microsoft.OpenApi/Reader/OpenApiModelFactory.cs:328]:
```suggestion
                await input.CopyToAsync(bufferStream, 1024, token).ConfigureAwait(false);
```

File: src/Microsoft.OpenApi/Reader/OpenApiModelFactory.cs [hunk 12]
```diff
@@ -193,146 +194,172 @@ public static T Parse<T>(string input,
...
                 }
-                catch (HttpRequestException ex)
+                else
                 {
-                    throw new InvalidOperationException($"Could not download the file at {url}", ex);
+                    // YAML or other non-JSON format; copy remaining input to a new stream.
+                    preparedStream = new MemoryStream();
+                    bufferStream.Position = 0;
+                    await bufferStream.CopyToAsync(preparedStream, 81920, token); // Copy buffered portion
+                    await input.CopyToAsync(preparedStream, 81920, token); // Copy remaining data
```
Comment (baywet) [inline src/Microsoft.OpenApi/Reader/OpenApiModelFactory.cs:345]:
```suggestion
                    await bufferStream.CopyToAsync(preparedStream, 81920, token).ConfigureAwait(false); // Copy buffered portion
                    await input.CopyToAsync(preparedStream, 81920, token).ConfigureAwait(false); // Copy remaining data
```

File: src/Microsoft.OpenApi/Reader/OpenApiModelFactory.cs [hunk 13]
```diff
@@ -193,146 +194,172 @@ public static T Parse<T>(string input,
...
-                        SecurityException or
-                        NotSupportedException)
+                format ??= InspectStreamFormat(input);
+
+                if (!format.Equals(OpenApiConstants.Json, StringComparison.OrdinalIgnoreCase))
                 {
-                    throw new InvalidOperationException($"Could not open the file at {url}", ex);
+                    // Buffer stream for non-JSON formats (e.g., YAML) since they require synchronous reading
+                    preparedStream = new MemoryStream();
+                    await input.CopyToAsync(preparedStream, 81920, token);
```
Comment (baywet) [inline src/Microsoft.OpenApi/Reader/OpenApiModelFactory.cs:357]:
```suggestion
                    await input.CopyToAsync(preparedStream, 81920, token).ConfigureAwait(false);
```

**Review by baywet — CHANGES_REQUESTED** [review]

File: src/Microsoft.OpenApi/Interfaces/IOpenApiReader.cs [hunk 14]
```diff
@@ -15,33 +15,40 @@ namespace Microsoft.OpenApi.Interfaces
...
         /// </summary>
         /// <param name="input">TextReader containing OpenAPI description to parse.</param>
         /// <param name="version">Version of the OpenAPI specification that the fragment conforms to.</param>
         /// <param name="diagnostic">Returns diagnostic object containing errors detected during parsing.</param>
         /// <param name="settings">The OpenApiReader settings.</param>
         /// <returns>Instance of newly created IOpenApiElement.</returns>
-        T ReadFragment<T>(TextReader input, OpenApiSpecVersion version, out OpenApiDiagnostic diagnostic, OpenApiReaderSettings settings = null) where T : IOpenApiElement;
+        T ReadFragment<T>(MemoryStream input, OpenApiSpecVersion version, out OpenApiDiagnostic diagnostic, OpenApiReaderSettings settings = null) where T : IOpenApiElement;
 
         /// <summary>
```
[the remark this answers is not in the document]
Comment (MaggieKimani1) [reply src/Microsoft.OpenApi/Interfaces/IOpenApiReader.cs:53]: Yeah we can abstract this from the interface considering it should be format-agnostic as users might need to register custom readers that don't implement JsonNodes

File: src/Microsoft.OpenApi/Reader/OpenApiModelFactory.cs [hunk 15]
```diff
@@ -118,12 +91,13 @@ public static async Task<ReadResult> LoadAsync(string url, OpenApiReaderSettings
...
+        /// <param name="token"></param>
         /// <returns>Instance of newly created IOpenApiElement.</returns>
         /// <returns>The OpenAPI element.</returns>
-        public static async Task<T> LoadAsync<T>(string url, OpenApiSpecVersion version, OpenApiReaderSettings settings = null) where T : IOpenApiElement
+        public static async Task<T> LoadAsync<T>(string url, OpenApiSpecVersion version, OpenApiReaderSettings settings = null, CancellationToken token = default) where T : IOpenApiElement
         {
-            var result = await RetrieveStreamAndFormatAsync(url).ConfigureAwait(false);
-            return Load<T>(result.Item1, version, out var _, result.Item2, settings);
+            var result = await RetrieveStreamAndFormatAsync(url, token).ConfigureAwait(false);
+            return await LoadAsync<T>(result.Item1, version, result.Item2, settings);
```
Comment (baywet) [inline src/Microsoft.OpenApi/Reader/OpenApiModelFactory.cs]:
```suggestion
            return await LoadAsync<T>(result.Item1, version, result.Item2, settings, token);
```

File: src/Microsoft.OpenApi/Reader/OpenApiModelFactory.cs [hunk 16]
```diff
@@ -159,6 +135,34 @@ public static async Task<ReadResult> LoadAsync(Stream input, string format = nul
...
+        /// <typeparam name="T"></typeparam>
+        /// <param name="input"></param>
+        /// <param name="version"></param>
+        /// <param name="format"></param>
+        /// <param name="settings"></param>
+        /// <returns></returns>
+        public static async Task<T> LoadAsync<T>(Stream input,
+                                                 OpenApiSpecVersion version,
+                                                 string format = null,
+                                                 OpenApiReaderSettings settings = null) where T : IOpenApiElement
```
Comment (baywet) [inline src/Microsoft.OpenApi/Reader/OpenApiModelFactory.cs:150]: missing cancellation token

**Review by baywet — CHANGES_REQUESTED** [review]

File: src/Microsoft.OpenApi/Interfaces/IOpenApiReader.cs (hunk 14 above)
[the remark this answers is not in the document]
Comment (baywet) [reply src/Microsoft.OpenApi/Interfaces/IOpenApiReader.cs]: ah right, I missing the agnostic aspect. Thanks for calling this out! Resolving.

File: src/Microsoft.OpenApi/Reader/OpenApiJsonReader.cs [hunk 17]
```diff
@@ -123,18 +110,55 @@ public async Task<ReadResult> ReadAsync(JsonNode jsonNode,
...
+            if (input is null) throw new ArgumentNullException(nameof(input));
+            if (settings is null) throw new ArgumentNullException(nameof(settings));
+
+            JsonNode jsonNode;
+            var diagnostic = new OpenApiDiagnostic();
+
+            // Parse the JSON text in the stream into JsonNodes
+            try
+            {
+                jsonNode = await JsonNode.ParseAsync(input, cancellationToken: cancellationToken).ConfigureAwait(false);
```
[the remark this answers is not in the document]
Comment (baywet) [reply src/Microsoft.OpenApi/Reader/OpenApiJsonReader.cs:133]: hallucination at its finest

File: src/Microsoft.OpenApi/Reader/OpenApiModelFactory.cs [hunk 18]
```diff
@@ -193,146 +199,179 @@ public static T Parse<T>(string input,
...
-            return null;
+            return (null, null);
+        }
+
+        private static string InspectInputFormat(string input)
+        {
+            return input.StartsWith("{", StringComparison.OrdinalIgnoreCase) || input.StartsWith("[", StringComparison.OrdinalIgnoreCase) ? OpenApiConstants.Json : OpenApiConstants.Yaml;
+        }
+
+        private static string InspectStreamFormat(Stream stream)
```
[the remark this answers is not in the document]
Comment (baywet) [reply src/Microsoft.OpenApi/Reader/OpenApiModelFactory.cs:311]: another hallucination

**Review by baywet — CHANGES_REQUESTED** [review]
Thank you for making the changes!

**Review by baywet — APPROVED** [review]
(clicked the wrong button 🤦)

**Pull request merged**